Buddies/Utilities/AlgoliaSearch.swift

```diff
@@ -0,0 +1,139 @@
+//
+//  AlgoliaSearch.swift
+//  Buddies
+//
+//  Created by Noah Allen on 2/12/19.
+//  Copyright © 2019 Jack and the Beans. All rights reserved.
+//
+
+/** Usage example:
+let search = AlgoliaSearch()
+let s = "2019/02/15".toDate()?.date
+let e = "2019/02/20".toDate()?.date
+search.searchActivities(startingAt: s, endingAt: e, atLocation: (41.1603988, -80.0866907)) { (activities: [String], err: Error?) in
+ 
+    if let error = err { print(error) }
+    print(activities)
+}
+**/
+
+import Foundation
+import InstantSearchClient
+
+protocol SearchIndex {
+    func search(_ query: Query, requestOptions: RequestOptions?, completionHandler: @escaping CompletionHandler) -> Operation
+}
+extension Index: SearchIndex {}
+
+protocol SearchClient {
+    func getIndex(withName indexName: String) -> SearchIndex
+}
+
+// Wrap the function so that we can use "SearchIndex" as a mock
+extension Client: SearchClient {
+    func getIndex(withName indexName: String) -> SearchIndex {
+        return index(withName: indexName)
+    }
+}
+
+class AlgoliaSearch {
+    static let ACTIVITY_INDEX = "BUD_ACTIVITIES"
+    var client: SearchClient
+
+    init (algoliaClient: SearchClient? = nil) {
+        self.client = algoliaClient != nil ? algoliaClient! : AlgoliaSearch.algoliaFactory()
+    }
+
+    // Constructs an instance of the Algolia Client SDK:
+    static func algoliaFactory() -> Client {
+        var algoliaAppID = "NOPE"
+        var algoliaApiKey = "NOPE"
+        // Grab API key from key file if it exists:
+        if let path = Bundle.main.path(forResource: "Keys", ofType: "plist") {
+            let keys = NSDictionary(contentsOfFile: path)
+            if
+                let k = keys,
+                let apiKey = k["AlgoliaSearchKey"] as? String,
+                let appid = k["AlgoliaAppId"] as? String {
+                algoliaAppID = appid
+                algoliaApiKey = apiKey
+            }
+        }
+        return Client(appID: algoliaAppID, apiKey: algoliaApiKey)
+    }
+
+    // Searches for Activities with optional params
+    func searchActivities(
+              withText: String? = nil,
+              matchingAnyTopicOf: [String]? = nil,
+              startingAt: Date? = nil,
+              endingAt: Date? = nil,
+              atLocation: (Double, Double)? = nil, // Tuple: (lat, lng)
+              upToDisatnce: Int = 20000, // In meters, defaults to 20km
+              usingIndex: SearchIndex? = nil,
```

Comment:
So callers need to understand indexes? Or is this just for unit testing?

Reply:
They do not _have_ to, but they _can_. It defaults to `nil`, in which case it will construct an instance of the activity index. It doesn't default to the actual index here because the call doesn't really fit into the space that a default param would.

Reply:
And to clarify, you can't reference `self` in the default param either, which we have to do to access the client.